baybe/insights/shap.py

```
@@ -0,0 +1,358 @@
+
+    Raises:
+        ValueError: If the provided background data set is empty.
+        TypeError: If the provided explainer class is incompatible with the surrogate.
+    """
+    if data.empty:
+        raise ValueError("The provided background data set is empty.")
+
+    if isinstance(explainer_cls, str):
+        explainer_cls = _get_explainer_cls(explainer_cls)
+
+    import torch
+
+    if use_comp_rep:
+
+        def model(x: np.ndarray) -> np.ndarray:
+            tensor = to_tensor(x)
+            with torch.no_grad():
+                output = surrogate._posterior_comp(tensor).mean
+            return output.numpy()
+
+    else:
+
+        def model(x: np.ndarray) -> np.ndarray:
+            df = pd.DataFrame(x, columns=data.columns)
+            with torch.no_grad():
+                output = surrogate.posterior(df).mean
+            return output.numpy()
+
+    # Handle special settings: Lime default mode is otherwise set to "classification"
+    kwargs = {"mode": "regression"} if explainer_cls.__name__ == "LimeTabular" else {}
+
+    try:
+        shap_explainer = explainer_cls(model, data, **kwargs)
+
+        # Explain first two data points to ensure that the explainer is working
+        if is_shap_explainer(shap_explainer):
+            shap_explainer(data.iloc[0:1])
+    except shap.utils._exceptions.InvalidModelError:
+        raise TypeError(
+            f"The selected explainer class {explainer_cls} does not support the "
+            f"provided surrogate model."
+        )
+    except TypeError as e:
+        if "not supported for the input types" in str(e) and not use_comp_rep:
+            raise NotImplementedError(
+                f"The selected explainer class {explainer_cls} does not support "
+                f"the experimental representation. Switch to computational "
+                f"representation or use a different explainer (e.g. the default "
+                f"shap.KernelExplainer)."
+            )
+        else:
```

Is this `else` necessary? I think it can be skipped since, right?

---

if its skipped and another `TypeError` than we expect (as identified by the string match) is raised we would silently drop it

---

I do not want to skip the `raise e` part but just the `else:`, since the code after the `if` won't be executed anyway if the `if` statement holds due to an exception being raised.

---

Obsolete, have removed the entire block and replaced it with proper input validation